### PR TITLE
Include all used C++ library headers in basic.h

### PR DESCRIPTION
**Pull request by Sumith1896**
Fixes https://github.com/symengine/symengine/issues/643
\cc @certik @isuruf

### CODE REVIEW

**Comment by certik** [comment]
I think if a C++ header file is only used in a `.cpp` file, then it should only be included there, since otherwise it is wasteful to include it in `.h` files. What do you think?

File: symengine/basic.h [hunk 1]
```diff
@@ -19,6 +19,17 @@
 #include <unordered_map>
 #include <cassert>
 #include <ciso646>
+#include <cmath>
+#include <complex>
+#include <stdlib.h>
```
Comment (isuruf) [inline symengine/basic.h]: should be changed to `<cstdlib>`

File: symengine/basic.h [hunk 2]
```diff
@@ -19,6 +19,17 @@
...
+#include <functional>
+#include <type_traits>
+#include <valarray>
+#include <iterator>
+#include <algorithm>
+#include <limits>
+#include <algorithm>
```
Comment (isuruf) [inline symengine/basic.h]: algorithm is included twice

File: symengine/basic.h [hunk 3]
```diff
@@ -19,6 +19,17 @@
 #include <unordered_map>
 #include <cassert>
 #include <ciso646>
+#include <cmath>
+#include <complex>
+#include <stdlib.h>
+#include <functional>
+#include <type_traits>
+#include <valarray>
```
Comment (isuruf) [inline symengine/basic.h]: This is only needed in a .cpp file.

**Comment by Sumith1896** [comment]
@certik Yes, valid point. Will take care.
@isuruf Thanks for the notes. I will look into that.

**Comment by Sumith1896** [comment]
@isuruf Could have a look at this now?

**Comment by certik** [comment]
Looks good to me.

**Comment by Sumith1896** [comment]
ping @isuruf

**Comment by certik** [comment]
I think that looks good. Merging.

**Comment by Sumith1896** [comment]
Thanks for the merge @certik

**Pull request merged**